### PR TITLE
TYP: Type ``MaskedArray.__new__``

### PR DESCRIPTION
**Pull request by MarcoGorelli**
~~This is work-in-progress~~

I'm trying to get the overloads to work when `data` is a subclass of `MaskedArray`, but am running into issues.

Essentially, if you pass a `MaskedArray` subclass to `MaskedArray`, then with `subok=True` (default) you get that subclass back, whereas with `subok=False` you just get `MaskedArray`
```python
In [5]: type(np.ma.MaskedArray(MySubclass([1,2,3])))
Out[5]: __main__.MySubclass

In [6]: type(np.ma.MaskedArray(MySubclass([1,2,3]), subok=False))
Out[6]: numpy.ma.MaskedArray
```

So, I'd like the test
```python
assert_type(np.ma.MaskedArray(MAR_subclass), MaskedArraySubclassC)
```
to pass. I've tried doing so by adding the overload
```python
    @overload
    def __new__(
        cls,
        data: _ArrayT,
        mask: _ArrayLikeBool_co = nomask,
        dtype: None = None,
        copy: bool = False,
        subok: Literal[True] = True,
        ndmin: int = 0,
        fill_value: _ScalarLike_co | None = None,
        keep_mask: bool = True,
        hard_mask: bool | None = None,
        shrink: bool = True,
        order: _OrderKACF | None = None,
    ) -> _ArrayT: ...
```

However, I then get
```
numpy/typing/tests/data/reveal/ma.pyi:445 - reveal mismatch:

    error:
    Expression is of type "MaskedArray[tuple[Any, ...], dtype[Any]]", not
    "MaskedArray[tuple[Any, ...], dtype[floating[_32Bit]]]"  [assert-type]
        assert_type(np.ma.MaskedArray(AR_f4), MaskedArray[np.float32])
        ^
_______________________________________________________________________________

numpy/typing/tests/data/reveal/ma.pyi:446 - reveal mismatch:

    error:
    Expression is of type "MaskedArray[tuple[Any, ...], dtype[Any]]", not
    "MaskedArraySubclass[complex128]"  [assert-type]
        assert_type(np.ma.MaskedArray(MAR_subclass), MaskedArraySubclassC)
        ^
```

Having said that, even without this extra overload, I think this is already an improvement on the (untyped) status-quo

### CODE REVIEW

**Comment by ranjodhsingh1729** [comment]
Hi,
I spent some time trying to understand this, and I may be missing something—but could you clarify what `subok` is supposed to do?
From what I can tell, the current behavior seems expected, but i might be wrong.

File: numpy/ma/core.pyi [hunk 1]
```diff
@@ -455,7 +462,68 @@ class MaskedIterator:
...
+        fill_value: _ScalarLike_co | None = None,
+        keep_mask: bool = True,
+        hard_mask: bool | None = None,
+        shrink: bool = True,
+        order: _OrderKACF | None = None,
+    ) -> _MaskedArray[_ScalarT]: ...
+    @overload
+    def __new__(
+        cls,
+        data: Any,
```
Comment (jorenham) [inline numpy/ma/core.pyi]: maybe `npt.ArrayLike` broad enough?

File: numpy/ma/core.pyi [hunk 2]
```diff
@@ -455,7 +462,68 @@ class MaskedIterator:
 
 class MaskedArray(ndarray[_ShapeT_co, _DTypeT_co]):
     __array_priority__: Any
-    def __new__(cls, data=..., mask=..., dtype=..., copy=..., subok=..., ndmin=..., fill_value=..., keep_mask=..., hard_mask=..., shrink=..., order=...): ...
+    @overload
+    def __new__(
+        cls,
+        data: _ArrayLike[_ScalarT],
+        mask: _ArrayLikeBool_co = nomask,
```
Comment (jorenham) [inline numpy/ma/core.pyi:462]: maybe I'm wrong here, but I thought integer masks also worked

File: numpy/ma/core.pyi [hunk 3]
```diff
@@ -455,7 +462,68 @@ class MaskedIterator:
...
+    ) -> _MaskedArray[_ScalarT]: ...
+    @overload
+    def __new__(
+        cls,
+        data: Any,
+        mask: _ArrayLikeBool_co,
+        dtype: _DTypeLike[_ScalarT],
+        copy: bool = False,
+        subok: bool = True,
+        ndmin: int = 0,
```
Comment (jorenham) [inline numpy/ma/core.pyi:481]: this returns an explicit shape type, so we can't have `ndim` interfere here:

```suggestion
        ndmin: Literal[0] = 0,
```

File: numpy/ma/core.pyi [hunk 4]
```diff
@@ -455,7 +462,68 @@ class MaskedIterator:
...
+        fill_value: _ScalarLike_co | None = None,
+        keep_mask: bool = True,
+        hard_mask: bool | None = None,
+        shrink: bool = True,
+        order: _OrderKACF | None = None,
+    ) -> MaskedArray[_ShapeT_co, dtype[_ScalarT]]: ...
```
Comment (jorenham) [inline numpy/ma/core.pyi]: what will `_ShapeT_co` be?

File: numpy/ma/core.pyi [hunk 5]
```diff
@@ -455,7 +462,68 @@ class MaskedIterator:
...
```
Comment (jorenham) [inline numpy/ma/core.pyi]: what's `_ShapeT_co`?

File: numpy/ma/core.pyi [hunk 6]
```diff
@@ -455,7 +462,68 @@ class MaskedIterator:
...
+    @overload
+    def __new__(
+        cls,
+        data: Any = None,
+        mask: _ArrayLikeBool_co = nomask,
+        dtype: DTypeLike = None,
```
Comment (jorenham) [inline numpy/ma/core.pyi]:
```suggestion
        dtype: DTypeLike | None = None,
```

File: numpy/ma/core.pyi [hunk 7]
```diff
@@ -455,7 +462,68 @@ class MaskedIterator:
...
+        fill_value: _ScalarLike_co | None = None,
+        keep_mask: bool = True,
+        hard_mask: bool | None = None,
+        shrink: bool = True,
+        order: _OrderKACF | None = None,
+    ) -> MaskedArray[_ShapeT_co, dtype[_ScalarT]]: ...
+    @overload
+    def __new__(
+        cls,
+        data: Any = None,
```
Comment (jorenham) [inline numpy/ma/core.pyi]: according to the docstring, `data` has no default:

```
An array class with possibly masked values.

Masked values of True exclude the corresponding element from any
computation.

Construction::

  x = MaskedArray(data, mask=nomask, dtype=None, copy=False, subok=True,
                  ndmin=0, fill_value=None, keep_mask=True, hard_mask=None,
                  shrink=True, order=None)

Parameters
----------
data : array_like
    Input data.
```

File: numpy/typing/tests/data/fail/ma.pyi [hunk 8]
```diff
@@ -143,3 +143,5 @@ MAR_td64 **= 2  # type: ignore[misc]
 MAR_1d_f8.swapaxes(axis1=1, axis2=0)  # type: ignore[call-arg]
 
 MAR_1d_f8.argsort(axis=(1,0))  # type: ignore[arg-type]
+
+np.ma.MaskedArray(np.array([1,2,3]), keep_mask='yes')  # type: ignore[call-overload]
```
Comment (jorenham) [inline numpy/typing/tests/data/fail/ma.pyi:147]: seems a bit arbitrary to only test this false positive case

File: numpy/typing/tests/data/reveal/ma.pyi [hunk 9]
```diff
@@ -1,14 +1,21 @@
...
+_ScalarT_co = TypeVar("_ScalarT_co", bound=generic, covariant=True)
 MaskedArray: TypeAlias = np.ma.MaskedArray[_AnyShape, dtype[_ScalarT]]
 _Array1D: TypeAlias = np.ndarray[tuple[int], np.dtype[_ScalarT]]
 
-class MaskedArraySubclass(MaskedArray[np.complex128]): ...
+class MaskedArraySubclass(MaskedArray[_ScalarT_co]): ...
+
+class IntoMaskedArraySubClass(Generic[_ScalarT_co]):
+    def __array__(self) -> MaskedArraySubclass[_ScalarT_co]:
+        ...
```
Comment (jorenham) [inline numpy/typing/tests/data/reveal/ma.pyi]:
```suggestion
    def __array__(self) -> MaskedArraySubclass[_ScalarT_co]: ...
```

(the `Into*` in the name makes me think that you know rust, is that correct :P ?)

**Comment by jorenham** [comment]
> I'm trying to get the overloads to work when `data` is a subclass of `MaskedArray`, but am running into issues.

Mypy or pyright? I know that mypy (still) does not have full support for `__new__`, and does not conform to the [typing spec](https://typing.python.org/en/latest/spec/constructors.html#new-method) in that regard. So if it's just mypy, and you can verify that pyright understand the additional overload, then you might as well leave it (the overload) in AFAIK.

**Comment by jorenham** [comment]
> I spent some time trying to understand this, and I may be missing something—but could you clarify what `subok` is supposed to do?

IMO, the `MaskedArray` docstring explains this pretty well:

```
subok : bool, optional
    Whether to return a subclass of `MaskedArray` if possible (True) or a
    plain `MaskedArray`. Default is True.
```

**Comment by ranjodhsingh1729** [comment]
Thanks for the reply. This one’s on me — probably just sleep-deprived.

File: numpy/ma/core.pyi (hunk 2 above)
Comment (MarcoGorelli) [reply numpy/ma/core.pyi:462]: This is the same annotation as in `__setmask__`

https://github.com/numpy/numpy/blob/8b2321d03c1abce671a15f9d08a97249bd2d161d/numpy/ma/core.pyi#L510

not sure if passing integer masks should be encouraged?

File: numpy/ma/core.pyi (hunk 5 above)
Comment (MarcoGorelli) [reply numpy/ma/core.pyi]: 🤦 thanks, what I wrote doesn't make sense

File: numpy/ma/core.pyi (hunk 3 above)
Comment (MarcoGorelli) [reply numpy/ma/core.pyi:481]: right, thanks - maybe fine to just type the dtype then, as is done for `array`?

https://github.com/numpy/numpy/blob/6eeaae25ddfacf586ee02ddc600618127fa79cfd/numpy/_core/multiarray.pyi#L474-L484

File: numpy/typing/tests/data/reveal/ma.pyi (hunk 9 above)
Comment (MarcoGorelli) [reply numpy/typing/tests/data/reveal/ma.pyi]: 😄 not an expert, but i put this together https://marcogorelli.github.io/polars-plugins-tutorial/

Comment (jorenham) [reply numpy/typing/tests/data/reveal/ma.pyi]: ohh nice; I was thinking about writing an [Lmo](https://github.com/jorenham/lmo/) plugin for polars actually, so I'll give this a read then :)

**Review by jorenham — APPROVED** [review]

**Comment by jorenham** [comment]
source of the new primer errors:
https://github.com/scikit-learn/scikit-learn/blob/da90c58fc3247adf147266106454d443b911b3a8/sklearn/model_selection/tests/test_search.py#L2945

It looks `ArrayLike` isn't broad enough, after all

**Review by jorenham — COMMENTED** [review]
small nit: for input types it's usually cleaner to use `object` instead of `Any` (in stubs)

File: numpy/ma/core.pyi (hunk 2 above)
Comment (jorenham) [reply numpy/ma/core.pyi:462]: > not sure if passing integer masks should be encouraged?

No idea to be honest. You probably know more about `np.ma` then I do, so your call.

File: numpy/ma/core.pyi (hunk 3 above)
Comment (jorenham) [reply numpy/ma/core.pyi:481]: Yea I think that's wise. Shape-typing could be nice to have here, but it'll probably  complicate things quite a bit.

File: numpy/typing/tests/data/reveal/ma.pyi [hunk 10]
```diff
@@ -1,14 +1,20 @@
-from typing import Any, Literal, TypeAlias, TypeVar, assert_type, NoReturn
+from typing import Any, Literal, TypeAlias, TypeVar, assert_type, NoReturn, Generic
```
Comment (jorenham) [inline numpy/typing/tests/data/reveal/ma.pyi]: nit

```suggestion
from typing import Any, Generic, Literal, NoReturn, TypeAlias, TypeVar, assert_type
```

File: numpy/ma/core.pyi (hunk 2 above)
Comment (MarcoGorelli) [reply numpy/ma/core.pyi:462]: 😄 not necessarily

tbh I'd be ok with punting on it for the sake of this issue, and if it needs updating, it can be updated throughout the file

**Comment by jorenham** [comment]
Thanks Marco :)
